sled-agent/src/zone_bundle.rs

```diff
@@ -1545,61 +1545,54 @@ async fn compute_bundle_utilization(
+// directory fails.
 async fn disk_usage(path: &Utf8PathBuf) -> Result<u64, BundleError> {
-    // Each OS implements slightly different `du` options.
-    //
-    // Linux and illumos support the "apparent" size in bytes, though using
-    // different options. macOS doesn't support bytes at all, and has a minimum
-    // block size of 512.
-    //
-    // We'll suffer the lower resolution on macOS, and get higher resolution on
-    // the others.
-    cfg_if::cfg_if! {
-        if #[cfg(target_os = "illumos")] {
-            const BLOCK_SIZE: u64 = 1;
-            const DU_ARG: &str = "-A";
-        } else if #[cfg(target_os = "linux")] {
-            const BLOCK_SIZE: u64 = 1;
-            const DU_ARG: &str = "-b";
-        } else if #[cfg(target_os = "macos")] {
-            const BLOCK_SIZE: u64 = 512;
-            const DU_ARG: &str = "-k";
-        } else {
-            compile_error!("unsupported target OS");
+    let path = path.clone();
+    // We could, alternatively, just implement this using `tokio::fs` to read
+    // the directory and so on. However, the `tokio::fs` APIs are basically just
+    // a wrapper around `spawn_blocking`-ing code that does the `std::fs`
+    // functions. So, putting the whole thing in one big `spawn_blocking`
+    // closure that just does all the blocking fs ops lets us spend less time
+    // creating and destroying tiny blocking tasks.
+    tokio::task::spawn_blocking(move || {
+        let mut sum = 0;
+        // Rather than recursing, we use a VecDeque as a queue of paths to scan
+        // next. This is so that if we're walking a very deeply nested directory
+        // structure, we don't overflow our stack.
+        //
+        // TODO(eliza): this could, conceivably, be parallelized by spawning a
+        // bunch of blocking tasks to walk each sub-directory, instead of using
+        // a queue. But, is optimizing this really all that important? It's
+        // probably far from the slowest part of zone bundle collection anyway...
+        let mut dirs_to_scan = std::collections::VecDeque::new();
+        dirs_to_scan.push_back(path.to_owned());
+
+        while let Some(dir) = dirs_to_scan.pop_front() {
+            let mk_dir_error = |err| BundleError::ReadDirectory {
+                directory: dir.clone(),
+                err,
+            };
+            for entry in std::fs::read_dir(&dir).map_err(mk_dir_error)? {
+                let entry = entry.map_err(mk_dir_error)?;
+                let mk_error = |err| BundleError::ReadDirectory {
+                    directory: Utf8PathBuf::try_from(entry.path())
+                        .unwrap_or_else(|_| path.clone()),
+                    err,
+                };
+                if entry.file_type().map_err(mk_error)?.is_dir() {
+                    let path = entry.path().try_into()?;
+                    dirs_to_scan.push_back(path);
+                } else {
+                    sum += entry.metadata().map_err(mk_error)?.len()
```

Comment:
Not sure if it's worth adding a new variant here, but this isn't really a `ReadDirectory` error anymore. It's something like a `StatError` or similar. Or we could add a context string to the error variant itself, if that's simpler?

Reply:
Yeah, I think there may also be some error variants that are only used by the `du` stuff...will look.

Reply:
Ah, nevermind the last bit, we also shell out to a bunch of zfs commands, so the `BundleError::Command` variants are still necessary.

Reply:
Oh, there's already a `BundleError::Metadata` variant, we could just use that!